### PR TITLE
Catch dimension mismatches between demos and policy

### PR DESCRIPTION
**Pull request by andrewcoh**
### Proposed change(s)

To address https://jira.unity3d.com/browse/MLA-928

I added unit tests for all the errors. I also checked this by modifying the CrawlerStatic/PushBlock action and observation spaces separately to make sure the correct errors were thrown.  This checks both continuous and discrete.

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)

https://jira.unity3d.com/browse/MLA-928

### Types of change(s)

- [x] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [x] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py [hunk 1]
```diff
@@ -106,8 +146,9 @@ def prepare_update(
     ) -> Dict[tf.Tensor, Any]:
         """
         Prepare inputs for update. .
-        :param mini_batch_demo: A mini batch of expert trajectories
-        :param mini_batch_policy: A mini batch of trajectories sampled from the current policy
+        :param policy: The policy learning from GAIL signal
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py:149]: Just randomly found this and fixed it. There are others that I can address separately.

Comment (chriselion) [reply ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py:149]: Wish there was a good way to automate checking this in CI, but don't think it's possible right now.

File: ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py [hunk 2]
```diff
@@ -42,6 +42,46 @@ def __init__(
             policy, 128, learning_rate, encoding_size, use_actions, use_vail
         )
         _, self.demonstration_buffer = demo_to_buffer(demo_path, policy.sequence_length)
+
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py:45]: I think we need to do this for BCModule too, right?
https://github.com/Unity-Technologies/ml-agents/blob/a6500a665e7d6a1a8f7cfb21d5d7147d3f61ca71/ml-agents/mlagents/trainers/components/bc/module.py#L41

Maybe better to move it to a utility, or pass the expected `BehaviorSpec` to `demo_to_buffer()`?

File: ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py [hunk 3]
```diff
@@ -42,6 +42,46 @@ def __init__(
...
+        # check number of visual observations in demmonstration match
+        for i in range(self.policy.vis_obs_size):
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py:47]: I think we need to check the sizes of the visual obs and vector obs too.

Comment (andrewcoh) [reply ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py:47]: Vector obs are checked below. I can check the size of the visual too.

File: ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py (hunk 2 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py:45]: good point. I'll think on it.

Comment (ervteng) [reply ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py:45]: Ah, this ties back into the fact that the trainers still use BrainParameters to describe behaviors. As a stopgap could we just create a function that compares two BrainParameters and convert the demo's BehaviorSpec to a BrainParameters object using `behavior_spec_to_brain_parameters`?

Comment (andrewcoh) [reply ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py]: @ervteng thats basically what I did manually. I can move it into the BrainParameters class

File: ml-agents/mlagents/trainers/demo_loader.py [hunk 4]
```diff
@@ -83,7 +83,7 @@ def make_demo_buffer(
 
 @timed
 def demo_to_buffer(
-    file_path: str, sequence_length: int
+    file_path: str, sequence_length: int, expected_brain_params: BrainParameters
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/demo_loader.py:86]: Would also make the expected_brain_params Optional and disable checks if None.

**Pull request merged**